pkg/operators/uidgidresolver/usergroupcache.go

```
@@ -0,0 +1,261 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uidgidresolver
+
+import (
+	"bufio"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/cachedmap"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/utils/host"
+)
+
+// UserGroupCache is a cache of user names, uids, group names and gids
+type UserGroupCache interface {
+	Start()
+	Stop()
+
+	GetUsername(uint32) string
+	GetGroupname(uint32) string
+}
+
+type userGroupCache struct {
+	userCache  cachedmap.CachedMap[uint32, string]
+	groupCache cachedmap.CachedMap[uint32, string]
+
+	loopFinished  chan struct{}
+	useCount      int
+	useCountMutex sync.Mutex
+	watcher       *fsnotify.Watcher
+}
+
+const (
+	passwdFileName = "passwd"
+	groupFileName  = "group"
+	baseDirPath    = "/etc"
+)
+
+var (
+	userGroupCacheSingleton *userGroupCache
+	userGroupErr            error
+	userGroupOnce           sync.Once
+
+	fullPasswdPath = filepath.Join(host.HostRoot, baseDirPath, passwdFileName)
+	fullGroupPath  = filepath.Join(host.HostRoot, baseDirPath, groupFileName)
+)
+
+func GetUserGroupCache() (UserGroupCache, error) {
+	userGroupOnce.Do(func() {
+		userGroupCacheSingleton, userGroupErr = newUserGroupCache()
+	})
+	return userGroupCacheSingleton, userGroupErr
+}
+
+func newUserGroupCache() (*userGroupCache, error) {
+	return &userGroupCache{
+		userCache:  cachedmap.NewCachedMap[uint32, string](2 * time.Second),
+		groupCache: cachedmap.NewCachedMap[uint32, string](2 * time.Second),
+	}, nil
+}
+
+func (cache *userGroupCache) Start() {
+	cache.useCountMutex.Lock()
+	defer cache.useCountMutex.Unlock()
+
+	// No uses before us, we are the first one
+	if cache.useCount == 0 {
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			log.Warnf("UserGroupCache: create watcher: %v", err)
```

Why doesn't it return an error? I'd rather return an error here and handle it on `PreGadgetRun() ` (by printing a warning if we can to ignore it)